clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -110,7 +110,8 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
      * @param latestLeaderEpoch latest leader's epoch.
      */
     void maybeUpdateLeaderEpoch(Optional<Integer> latestLeaderEpoch) {
-        if (!currentLeaderEpoch.equals(latestLeaderEpoch)) {
+        if (latestLeaderEpoch.isPresent()
```

Comment:
Hmm, this looks like a change in behavior. Previously we would override a known current leader epoch if `latestLeaderEpoch` is not defined. I am not sure if that was intentional. I recall there we had some logic which assumed that metadata version might be downgraded and no longer provide a leader epoch. Not sure it matters here though since we are not changing the leader, just updating the epoch.

Reply:
The change is intentional. The optimisation that KIP-951 proposed is that producer-batch should skip backoff period if "new leader" is known. This change corrects the previous logic to update the batch's leader-epoch only if its newer. I added a test in `ProducerBatchTest.java` to reflect that.